Allow activity change on start of switching activity

ehendrix23 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Provide user with the option to update current activity when the switch to a new activity is initiated instead of when the switch is completed.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [X] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #307111
- This PR is related to issue: 
- Link to documentation pull request: TBD

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

ehendrix23 committed:
Allow activity change on start of switch

Allow activity to be updated when a switch to a new activity is initiated instead of when it is completed.

ehendrix23 commented:
@bdraco,
You added the config UI entries for Harmony. I've been trying to add the toggle to allow the user to switch between setting the new activity upon start of the switching the activity or upon finishing it.
Default is when it is finished as that is how it always was.
But when I try it out I just do not see the additional option in the UI. So I am doing something wrong or missing something. 
Appreciate your help on it.

bdraco commented on homeassistant/components/harmony/translations/nl.json:
```diff
@@ -27,7 +27,8 @@
                     "activity": "De standaardactiviteit die moet worden uitgevoerd wanneer er geen is opgegeven.",
-                    "delay_secs": "De vertraging tussen het verzenden van opdrachten."
+                    "delay_secs": "De vertraging tussen het verzenden van opdrachten.",
+                    "activity_notify": "Verander de activiteit bij het starten van de activiteits verandering."
```
No need to add this here as it will get overwritten by lokalise.  Only `strings.json` and `en.json` should be in the PR

ehendrix23 replied:
And here I was trying to be helpful. :-) 
Reverted.

bdraco commented on homeassistant/components/harmony/remote.py:
```diff
@@ -207,6 +217,7 @@ def device_info(self):
             ),
             "name": self.name,
             "model": model,
+            "protocol": self._client.protocol,
```
https://developers.home-assistant.io/docs/device_registry_index/#device-properties

Protocol isn't a valid key here

ehendrix23 replied:
Ah, that's where it is used for. Removed. :-)

bdraco commented on homeassistant/components/harmony/remote.py:
```diff
@@ -57,6 +55,7 @@
         vol.Optional(ATTR_ACTIVITY): cv.string,
         vol.Required(CONF_NAME): cv.string,
         vol.Optional(ATTR_DELAY_SECS, default=DEFAULT_DELAY_SECS): vol.Coerce(float),
+        vol.Optional(ATTR_ACTIVITY_NOTIFY, default=False): cv.boolean,
```
Since this is converted to a config flow, we don't update the legacy schema with new options anymore. We should take this out.

ehendrix23 replied:
Removed that 1 line, should the rest for PLATFORM_SCHEMA stay in there or be removed as well?

bdraco replied:
We probably still have people who haven’t upgraded yet so best to leave it in for now

bdraco commented on homeassistant/components/harmony/config_flow.py:
```diff
@@ -190,6 +192,10 @@ async def async_step_init(self, user_input=None):
                 vol.Optional(
                     ATTR_ACTIVITY, default=self.config_entry.options.get(ATTR_ACTIVITY),
                 ): vol.In(remote.activity_names),
+                vol.Optional(
```
We will need to add a test for this.

bdraco commented:
> @bdraco,

> But when I try it out I just do not see the additional option in the UI. So I am doing something wrong or missing something.
> Appreciate your help on it.

I'm taking a peek now

bdraco commented on homeassistant/components/harmony/__init__.py:
```diff
@@ -67,7 +74,7 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
     options = dict(entry.options)
     modified = 0
-    for importable_option in [ATTR_ACTIVITY, ATTR_DELAY_SECS]:
+    for importable_option in [ATTR_ACTIVITY, ATTR_DELAY_SECS, ATTR_ACTIVITY_NOTIFY]:
```
Since we don't update legacy platform schema, this can be removed here since it will never be imported

ehendrix23 replied:
Same here, removed the item I added but figuring leaving rest in there for folks coming from older HA's?

bdraco replied:
> Same here, removed the item I added but figuring leaving rest in there for folks coming from older HA's?

Exactly

bdraco commented:
Is there a branch for aioharmony 0.2.5 ?

bdraco commented:
It appears to be working as expected after I removed the all the references to `ClientCallbackType` in `remote.py` since I didn't have version 0.2.5

<img width="413" alt="Screen Shot 2020-06-11 at 9 36 50 PM" src="https://user-images.githubusercontent.com/663432/84458864-96b16200-ac2b-11ea-9182-023126b7ee51.png">

..and the config entry is updated in `.storage/core.config_entries`

```
            {
                "connection_class": "local_push",
                "data": {
                    "host": "192.168.209.238",
                    "name": "Harmony Hub"
                },
                "domain": "harmony",
                "entry_id": "4dff1ccd0c77489cba876d11f921e511",
                "options": {
                    "activity": "Watch Apple TV",
                    "activity_notify": true,
                    "delay_secs": 0.4
                },
                "source": "ssdp",
                "system_options": {
                    "disable_new_entities": false
                },
                "title": "Harmony Hub",
                "unique_id": "xxxxx",
                "version": 1
            },
```

bdraco commented on homeassistant/components/harmony/remote.py:
```diff
@@ -166,12 +168,20 @@ async def async_added_to_hass(self):
-            connect=self.got_connected,
-            disconnect=self.got_disconnected,
-        )
+        if self._activity_notify:
+            self._client.callbacks = ClientCallbackType(
+                new_activity_starting=self.new_activity,
+                config_updated=self.new_config,
+                connect=self.got_connected,
+                disconnect=self.got_disconnected,
+            )
+        else:
+            self._client.callbacks = ClientCallbackType(
+                new_activity=self.new_activity,
+                config_updated=self.new_config,
+                connect=self.got_connected,
+                disconnect=self.got_disconnected,
+            )
```
```suggestion
        callbacks = {
           config_updated: self.new_config,
           connect: self.got_connected
           disconnect: self.got_disconnected
        }
        if self._activity_notify:
            callbacks["new_activity_starting"] = self.new_activity
        else:
             callbacks["new_activity"] = self.new_activity
        self._client.callbacks = ClientCallbackType(**callbacks)
```

Doesn't save that many lines but maybe something the above?

ehendrix23 replied:
> Is there a branch for aioharmony 0.2.5 ?

Yeah: 

> Is there a branch for aioharmony 0.2.5 ?

Yep, still need to finish some more testing on it:
https://github.com/ehendrix23/aioharmony/archive/dev.zip

ehendrix23 replied:
Yep, looks cleaner. :-) Updated.

bdraco commented:
@ehendrix23 Nice job on this.  Seems like this will significantly improve responsiveness of activity changes.  I think it also implements #30711 if I'm not mistaken.

ehendrix23 commented:
> @ehendrix23 Nice job on this. Seems like this will significantly improve responsiveness of activity changes. I think it also implements #30711 if I'm not mistaken.

Yep, that will implement it. I was waiting for slixmpp to be fixed so it can work with HA again as my newer version that supports XMPP required it.
Now that it is fixed I can do updates on the latest version knowing it will end up in HA. :-)

ehendrix23 commented:
> It appears to be working as expected after I removed the all the references to 

OK, so what am I missing? I have HASS core for development and then I have my HASS that is running for my house.
I tried making harmony a custom component so I can use updated version (something I've done in the past) but config UI is still showing as is.

Are there any additional steps, scripts, ... to update HASS for the UI? It did not look so to me.

ehendrix23 committed:
Updates based on feedback

Some items are not required to be done as YAML is not used anymore.
Cleaned-up some code.

bdraco commented:
> > It appears to be working as expected after I removed the all the references to
> 
> OK, so what am I missing? I have HASS core for development and then I have my HASS that is running for my house.
> I tried making harmony a custom component so I can use updated version (something I've done in the past) but config UI is still showing as is.
> 
> Are there any additional steps, scripts, ... to update HASS for the UI? It did not look so to me.

I usually rsync the files into my test instance like this to make sure I get everything

`rsync -av /volume1/docker/dev_root/home-assistant/homeassistant/components/harmony/ custom_components/harmony/`

And then restart, that's it.

ehendrix23 commented:
> > > It appears to be working as expected after I removed the all the references to
> > 
> > 
> > OK, so what am I missing? I have HASS core for development and then I have my HASS that is running for my house.
> > I tried making harmony a custom component so I can use updated version (something I've done in the past) but config UI is still showing as is.
> > Are there any additional steps, scripts, ... to update HASS for the UI? It did not look so to me.
> 
> I usually rsync the files into my test instance like this to make sure I get everything
> 
> `rsync -av /volume1/docker/dev_root/home-assistant/homeassistant/components/harmony/ custom_components/harmony/`
> 
> And then restart, that's it.

Got it working. :-) I had renamed my previous folder in there to harmony_old and I think it must have been picking that up.
Once I removed it and restarted it complained it could not find version 0.2.5 for aioharmony. :-)
Now I can see it in the GUI with the default of turned off so that existing behavior stays. :-)

Will do more testing later today ... need to do some paying work 1st. :-)

ehendrix23 committed:
Fix for change on how to set callbacks

How callbacks are set now one has to set the new_activity and new_activity_starting as well, even just with None.

ehendrix23 committed:
Added callback update

Added so that when it is changed in the UI the callbacks will be changed as well.

ehendrix23 committed:
Added test cases for notify setting

Added test cases for config flow to test new setting for activity notifications.

ehendrix23 commented:
@bdraco , aioharmony 0.2.5 is released now. 
I added the test and executed confirming it is working.

bdraco commented:
I'll give it a spin this weekend

bdraco commented:
I have an automation that triggers a relay tied to a esphome board that triggers a TV lift when the harmony is turned on.  It used to have ~25s delay to bring down the TV.  Now it starts within a second of hitting the activity on the harmony with the option turned on.

Nice job on this 👍
